content/post/2019-12-19-mutable-api.Rmarkdown

```diff
@@ -0,0 +1,184 @@
+
+Let's create our own replacement function to make sure we got it right!
+
+```{r}
+x <- 1:5
+x
+
+# function that will replace all values of x
+# with the new value
+`replace_all<-` <- function(x, value) {
+  x[seq_along(x)] <- value
+  x
+}
+
+# the argument called value is passed at the right of the arrow
+replace_all(x) <- 42
+x
+```
+
+So we've modified x, but not in place, see below the same code with `tracemem()`
+
+```{r}
+x <- 1:5
+tracemem(x)
+`replace_all<-` <- function(x, value) {
+  x[seq_along(x)] <- value
+  x
+}
+replace_all(x) <- 42
+x
+```
+
+So replacement functions are a standard way to give a mutable flavour to R code. Let's move on to another mutable feel.
+
+## Exposing the C API in xml2
+
+With `xml2` you can modify and remove XML nodes from a tree which makes you feel the tree is mutable.
+
+See for instance the code in [our blog post about READMEs](/2019/12/03/readmes/#other-size-indicators)
+
+```r
+xml2::xml_replace(xml2::xml_find_all(xml, "//softbreak"),
+                      xml2::read_xml("<text>\n</text>"))
+```
+
+That code changes nodes in the `xml` object without our assigning it back to it.
+
+The reasons `xml2` works this way is that the package is a binding to [the C libxml2 API](http://xmlsoft.org/), where you are supposed to manage memory allocation manually. [`xml2` does handle memory management for you](https://xml2.r-lib.org/#compared-to-the-xml-package) but [`xml2` docs explain how one should be careful when removing nodes because of possible memory issues](https://xml2.r-lib.org/articles/modification.html#removing-nodes). 
+
+Maybe XML data in itself is unusual for you, and maybe the behaviour above is even more unusual for you, but it's a handy package, [if only to tinker with READMEs](/2019/12/03/readmes/#other-size-indicators). :grin:
+
+## Interfacing an external process that's actually mutable in ps::ps_handle()
+
+Now, speaking of objects that are actually mutable, the [`ps` package](http://ps.r-lib.org/) offers an interesting example: the [`ps_handle()`](http://ps.r-lib.org/reference/ps_handle.html) function creates an object that's essentially a pointer to a system process. System processes are of course mutable, they run, then die, can be suspended, etc.
+
+In the example below we launch a system call using `processx`, create a `ps_handle` object corresponding to it i.e. just an external pointer with an S3 class, and we query its status using `ps`. 
+
+```{r}
+p <- processx::process$new("sleep", "5")
+```
+
+With such a definition, after 5 seconds the process will no longer exist. 
+
+```{r, error=TRUE}
+p$get_pid()
+phandle <- ps::ps_handle(p$get_pid())
```

Comment:
It is better to use `p$as_ps_handle()`. I would just print `p` and also `phandle`, to show that pids are the same. I would also add a paragraph that explains how the external ptr in processx is different from the one in ps and why there are two packages for this.